[Order Details] Refactor order detail rows to account for async operations

iamgabrielma opened the pull request:
<!-- Remember about a good descriptive title. -->

Closes: WOOMOB-424

## Description
This PR builds from the idea of https://github.com/woocommerce/woocommerce-ios/commit/2cc6ceb53c30e40915d0a3743c19627d0a4dbf18 of not assuming that all order detail rows will be built immediately. We refactor how we create those by separating into static and payments sections for now, wrap everything in a task, and only return when all sections are ready. 

This is pre-work to be able to use async-await for displaying failed receipts ([#](https://github.com/woocommerce/woocommerce-ios/pull/15558)), as well as a general improvement for future cases where rows might depend on network requests.

## Testing information

Smoke test order creation & order details, check that we can process a payment correctly, and see that dynamic rows are generated when needed (eg: shipping label button for physical products, receipts row after a card payment transaction, ...).

I sprinkled a few of main actors around around payment section creation, as otherwise we would crash when performing actions like generating shipping labels (as accesses storage and dispatches actions internally), as well as when checking receipt eligibility. I've also injected a new dependency for receipt eligibility, otherwise unit tests will be blocked due leaking continuations.

I had some troubles with a shipping labels [test](https://github.com/woocommerce/woocommerce-ios/pull/15588/commits/c2b5c9bc9027aff380d918e9399bfd817b899c80), and confirmed with @itsmeichigo that was [missing a flag](https://github.com/woocommerce/woocommerce-ios/pull/15588/commits/2a081cf87db41ce08e4b08baccbfebf1b7e0fa4c), which made it never fail CI but fail locally because running in debug.

## Screenshots
_Too big for GH: https://indiemelon.mystagingwebsite.com/wp-content/uploads/2025/05/Simulator-Screen-Recording-iPhone-16-US-store-2025-05-02-at-13.46.17.gif_

---

## Reviewer (or Author, in the case of optional code reviews):

Please make sure these conditions are met before approving the PR, or request changes if the PR needs improvement:

- [ ] The PR is small and has a clear, single focus, or a valid explanation is provided in the description. If needed, please request to split it into smaller PRs.
- [ ] Ensure Adequate Unit Test Coverage: The changes are reasonably covered by unit tests or an explanation is provided in the PR description.
- [ ] Manual Testing: The author listed all the tests they ran, including smoke tests when needed (e.g., for refactorings). The reviewer confirmed that the PR works as expected on all devices (phone/tablet) and no regressions are added.

iamgabrielma committed:
cherry pick 2cc6ceb

iamgabrielma committed:
extract finishSection

iamgabrielma committed:
Separate sections into static v async (payment)

We need to mark both the Task and the payment methods to run on the Main Actor, otherwise we’ll see crashes due to:

- Injecting products into shippingLabelOrderItemsAggregator, this is fetched from storage
- createPaymentSection calls isEligibleForBackendReceipt, which dispatches an action to the store and must be ran in main thread.

iamgabrielma committed:
lint

iamgabrielma committed:
restore unnecessary changes

iamgabrielma committed:
restore more unneccesary refactor

iamgabrielma committed:
Make section builder truly async by removing the task wrapper

await dataSource.reloadSections does not await anything, it just schedules a task to run in the background but doesn’t wait for it

iamgabrielma committed:
Inject ReceiptEligibilityUseCaseProtocol

Tests lock out because we’re leaking the continuation on receipts

iamgabrielma committed:
Render payment section specifically on its place

Due being async, it would always be appended at the end, rather than on a specific place within order details

iamgabrielma committed:
Skip test temporarily

This test seems to be incorrect, asserts for nil value but checks for shouldShowShippingLabelCreation, which is true in this case as is set in the test, but it doesn’t check whether the order already has non-refunded shipping labels.

iamgabrielma committed:
lint

iamgabrielma committed:
trigger reload sections

iamgabrielma committed:
Fix test

`shouldShowShippingLabelCreation` always returns true if `revampedShippingLabelCreation` feature flag is enabled. It didn’t fail CI because of production build, but seems this test was always failing in debug.

iamgabrielma committed:
delete debug comments

jaclync commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsViewModel.swift:
```diff
@@ -126,7 +126,9 @@ final class OrderDetailsViewModel {
+            Task { @MainActor in
+                await dataSource.reloadSections()
+            }
             onUIReloadRequired?()
```
❓ Shall this be called in the Task above after `dataSource.reloadSections`? Just to not assume the implementation that it calls `dataSource.onUIReloadRequired`, what about removing the `onUIReloadRequired` call at the end of `dataSource.reloadSections`?

iamgabrielma replied:
I'm a bit torn about this one in the context of this PR. It's true that we're duplicating the call to `onUIReloadRequired`, which is not necessary as we're already doing so internally when reloading sections.

Also agree that we should remove the invokation at the end of `dataSource.reloadSections`, we shouldn't assume the implementation of `reloadSections` contains the call to `onUIReloadRequired`. The problem is...we use this assumption in multiple places, and a handful of view controllers reach directly to `dataSource.reloadSections`, which I don't think they should, but changing this would involve going further away from the initial issue.

For the moment I've removed the duplication here https://github.com/woocommerce/woocommerce-ios/pull/15588/commits/c4c8595e123127e74d5f4ef801b38636f3012f94 just because it happens within order notes only, leaving the previous reliance on the data source as it is.

I'll timebox some time in this week's backlog handling just to see if it's a quick fix or we need to do deeper changes 👍

jaclync commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift:
```diff
@@ -2033,3 +2050,15 @@ extension OrderDetailsDataSource {
+        var updatedRows = rows
+        if isEligibleForRefund {
+            updatedRows.append(.issueRefundButton)
+        }
```
❓ Any reasons this refund button row isn't part of the rows logic in `createPaymentSection`? Could `createPaymentSection` just return the `Section` with all the rows there?

iamgabrielma replied:
No reason, must been part of my previous attempts. Updated! https://github.com/woocommerce/woocommerce-ios/pull/15588/commits/439a104f65dc0a9c1a7208515de93bbbe278c234

jaclync commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift:
```diff
@@ -1480,28 +1440,85 @@ extension OrderDetailsDataSource {
+    private func isEligibleForBackendReceipt() async -> Bool {
+        guard !isEligibleForPayment else { return false }
+        return await withCheckedContinuation { continuation in
+            receiptEligibilityUseCase.isEligibleForBackendReceipts { isEligible in
```
In the future, `isEligibleForBackendReceipts` can also have an async/await version.

jaclync commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift:
```diff
@@ -1312,69 +1336,6 @@ extension OrderDetailsDataSource {
             return Section(category: .shippingLines, title: Title.shippingLines, rows: Array(repeating: .shippingLine, count: shippingLines.count))
         }()
 
-        let customerInformation: Section? = {
-            var rows: [Row] = []
-
-            /// Customer Note
-            /// Always visible to allow adding & editing.
-            rows.append(.customerNote)
-
-            /// Shipping Address
-            /// Almost always visible to allow editing.
-            let orderContainsOnlyVirtualProducts = self.products.filter { (product) -> Bool in
-                return items.first(where: { $0.productID == product.productID}) != nil
-            }.allSatisfy { $0.virtual == true }
-
-
-            if order.shippingAddress != nil && orderContainsOnlyVirtualProducts == false {
-                rows.append(.shippingAddress)
-            }
-
-            /// Billing Address
-            /// Always visible to allow editing.
-            rows.append(.billingDetail)
-
-            /// Return `nil` if there is no rows to display.
-            guard rows.isNotEmpty else {
-                return nil
-            }
-
-            return Section(category: .customerInformation, title: Title.information, rows: rows)
-        }()
```
nit: just for easier review in the future, improvements to the code like moving unchanged function around / indentation etc. that aren't directly related to the PR could be skipped to minimize the diffs. This way, the focus is clear from reading the changes, especially for a bigger PR. Like for the move of `customerInformation`, it required some manual check to know that this is a change of location without code changes.

iamgabrielma replied:
I totally agree, I did a bunch of cleanup to avoid this in previous commits but for some reason I left this one here. Sorry for that!

jaclync commented on WooCommerce/Classes/ViewModels/Order Details/OrderDetailsDataSource.swift:
```diff
@@ -1185,14 +1192,33 @@ extension OrderDetailsDataSource {
+        self.sections = sections
+
+        self.updateOrderNoteAsyncDictionary(orderNotes: orderNotes)
+        self.onUIReloadRequired?()
```
As in the comment in `OrderDetailsViewModel`, this call could be replaced by the call in `OrderDetailsViewModel` just to maintain the pre-existing behavior.

iamgabrielma replied:
Yes, I don't like to have this hidden in the implementation details either. Having an UI reload closure to be invoked in the data source seems wrong. As with https://github.com/woocommerce/woocommerce-ios/pull/15588#discussion_r2074762253 I'll take a look to it separately just to keep things separated and test changes throughfully.

jaclync reviewed (approved):
Thanks for making the changes to make the reloadSections support async properly! Order details loaded as expected, tested with completed orders with receipt, failed orders, and orders with shipping labels. :shipit:

iamgabrielma committed:
remove unnecessary extension

iamgabrielma committed:
remove unnecessary

the UI reload invokation is already called inside reloadSections, no need for a second call outside the task
